crates/sui-core/src/authority/authority_store.rs

```diff
@@ -259,6 +297,51 @@ impl<const ALL_OBJ_VER: bool, S: Eq + Serialize + for<'de> Deserialize<'de>>
         self.executed_sequence.insert(&seq, digest).unwrap();
     }
 
+    /// Add a number of certificates to the pending transactions as well as the
+    /// certificates structure if they are not already executed.
+    pub fn add_pending_certificates(
+        &self,
+        certs: Vec<(TransactionDigest, CertifiedTransaction)>,
+    ) -> SuiResult<()> {
+        let first_index = self
+            .next_pending_seq
+            .fetch_add(certs.len() as u64, std::sync::atomic::Ordering::Relaxed);
+
+        let batch = self.pending_execution.batch();
+        let batch = batch.insert_batch(
+            &self.pending_execution,
+            certs
+                .iter()
+                .enumerate()
+                .map(|(num, (digest, _))| ((num as u64) + first_index, digest)),
```

Comment:
Based on this, we can't call this function concurrently from multiple tasks. Is this documented or enforced anywhere?

I think it might be good to start adding checks when we expect a function not to run concurrently, e.g.

```
let guard = if let Some(g) = self.concurrency_guard.try_lock() {
    g
} else {
    warn!("function was called concurrently");
    debug_assert!(false);
    self.concurrency_guard.lock();
}
```

It's getting hard to remember all the places we assume this constraint.

Reply:
In this instance I think we can run this concurrently, I tried to document why:
```
    /// This function may be run concurrently: it increases atomically an internal index
    /// by the number of certificates passed, and then records the certificates and their
    /// index. If two instanced run concurrently, the indexes are guaranteed to not overlap
    /// although some certificates may be included twice in the `pending_execution`, and 
    /// the same certificate may be written twice (but that is OK since it is valid.)
```

Reply:
Ah, ok, I missed that. Thanks!